apps/opencs/view/render/terrainselection.cpp

```diff
@@ -0,0 +1,251 @@
+#include <osg/Geometry>
+#include <osg/PositionAttitudeTransform>
+
+#include <components/esm/loadland.hpp>
+
+#include "../../model/world/cellcoordinates.hpp"
+#include "../../model/world/columnimp.hpp"
+#include "../../model/world/idtable.hpp"
+
+#include "worldspacewidget.hpp"
+
+namespace
+{
+    const int cellSize {ESM::Land::REAL_SIZE};
+    const int landSize {ESM::Land::LAND_SIZE};
+    const int landTextureSize {ESM::Land::LAND_TEXTURE_SIZE};
+}
+
+CSVRender::TerrainSelection::TerrainSelection(osg::Group* parentNode, WorldspaceWidget *worldspaceWidget):
+mParentNode(parentNode), mWorldspaceWidget (worldspaceWidget)
+{
+    osg::ref_ptr<osg::Geometry> geometry = new osg::Geometry();
+
+    // Color
+    osg::ref_ptr<osg::Vec4Array> colors = new osg::Vec4Array{};
+    colors->push_back(osg::Vec4f(0.f, 0.5f, 0.f, 1.f));
+
+    geometry->setColorArray(colors, osg::Array::Binding::BIND_OVERALL);
+
+    // Primitive
+    osg::ref_ptr<osg::DrawArrays> drawArrays = new osg::DrawArrays{osg::PrimitiveSet::Mode::LINES};
+
+    geometry->addPrimitiveSet(drawArrays);
+    geometry->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::Values::OFF);
+
+    mGeode = new osg::Geode();
+    mGeode->addDrawable(geometry);
+
+    activate();
+}
+
+CSVRender::TerrainSelection::~TerrainSelection()
+{
+    deactivate();
+}
+
+std::vector<std::pair<int, int>> CSVRender::TerrainSelection::getTerrainSelection() const
+{
+    return mSelection;
+}
+
+void CSVRender::TerrainSelection::selectTerrainTexture(const WorldspaceHitResult& hit)
+{
+    mSelection.clear();
+
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+
+    if (std::find(mSelection.begin(), mSelection.end(), localPos) == mSelection.end())
+        mSelection.push_back(localPos);
+
+    update();
+}
+
+void CSVRender::TerrainSelection::onlyAddSelect(const WorldspaceHitResult& hit)
+{
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+
+    if (std::find(mSelection.begin(), mSelection.end(), localPos) == mSelection.end())
+            mSelection.push_back(localPos);
+
+    update();
+}
+
+void CSVRender::TerrainSelection::toggleSelect(const WorldspaceHitResult& hit)
+{
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+    const auto iter = std::find(mSelection.begin(), mSelection.end(), localPos);
+
+    if (iter != mSelection.end()) {
+        mSelection.erase(iter);
+    }
+    else {
+        mSelection.push_back(localPos);
+    }
+    update();
+}
+
+void CSVRender::TerrainSelection::activate()
+{
+    mParentNode->addChild(mGeode);
+}
+
+void CSVRender::TerrainSelection::deactivate()
+{
+    mParentNode->removeChild(mGeode);
+}
+
+std::pair<int, int> CSVRender::TerrainSelection::toTextureCoords(osg::Vec3d worldPos) const
+{
+    const double xd {worldPos.x() * landTextureSize / cellSize+(landTextureSize / cellSize)/4};
+    const double yd {worldPos.y() * landTextureSize / cellSize+(landTextureSize / cellSize)/4};
```

Comment:
It does not look right too.
Technically, you use something like this:
```
xd = landTextureSize / cellSize * (worldPos.x() + 0.25f)
```
I wonder why you need to add the 0.25 to world coordinate.

Reply:
I think the original author mentioned something about needing this offset to make texture tiles line up. But I could be remembering incorrectly. It does look odd. Might be a symptom of a problem elsewhere?

Reply:
I haven't figured out a good implementation for handling texture selection and coordinates. This might not be a good way to do it. The offset is handled in another place, this "/ 4", if I remember correctly, is because vertex coordinates (land geometry) and texture coordinates (land textures) are sized in ratio of 1:4 (16 textures to 64 (+ 1 shared) vertexes).

Reply:
Sorry, zinnschlags explanation was the correct one. Texture nudging is done somewhere by -0.25f and +0.25f, and this needs to be accounted here. Vertex coordinate calculations probably shouldn't be nudged though. (fixed this in latest commit).